### PR TITLE
Add registrable DiscoverActions to discover page

### PR DESCRIPTION
**Pull request by gaobinlong**
### Description

This PR adds a registrable DiscoverActions component to DiscoverCanvas in the discover page, so that other plugins can register their own actions, and then display the actions in the Discover page.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7786

## Screenshot

image

## Testing the changes

If no actions are registered, then no impact on the discover page.
All of the registered actions will be associated with their own feature flags, if users don't enable the feature flags, then no actions will be shown in the discover page.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data_explorer/public/plugin.ts [hunk 1]
```diff
@@ -43,11 +44,12 @@ export class DataExplorerPlugin
   private appStateUpdater = new BehaviorSubject<AppUpdater>(() => ({}));
   private stopUrlTracking?: () => void;
   private currentHistory?: ScopedHistory;
+  private discoverActions: DiscoverAction[] = [];
 
   public setup(
     core: CoreSetup<DataExplorerPluginStartDependencies, DataExplorerPluginStart>,
     { data }: DataExplorerPluginSetupDependencies
-  ): DataExplorerPluginSetup {
+  ) {
```
Comment (SuZhou-Joe) [inline src/plugins/data_explorer/public/plugin.ts:52]: Seems unnecessary change.

File: src/plugins/data_explorer/public/plugin.ts [hunk 2]
```diff
@@ -105,6 +107,7 @@ export class DataExplorerPlugin
             ...withNotifyOnErrors(coreStart.notifications.toasts),
           }),
           viewRegistry: viewService.start(),
+          discoverActions: this.discoverActions,
```
Comment (SuZhou-Joe) [inline src/plugins/data_explorer/public/plugin.ts:110]: Nit: maybe `actions` is good enough as we are already under the context of discover.

**Comment by SuZhou-Joe** [comment]
Do not know if we have taken a look on the existing features of ui_actions_service.

File: src/plugins/data_explorer/public/plugin.ts (hunk 1 above)
Comment (gaobinlong) [reply src/plugins/data_explorer/public/plugin.ts]: Changed that, thanks!

**Comment by gaobinlong** [comment]
> Do not know if we have taken a look on the existing features of ui_actions_service.
> 
> https://github.com/opensearch-project/OpenSearch-> Dashboards/blob/cb273364086057b542a55d2bbe28631313a22df9/src/plugins/ui_actions/README.m

Yeah, we have considered that solution but it doesn't meet our requirement, ui actions only supports a menu button with multiple actions, the style is different from the UI mockup for `generate anomaly detector` and `text to visualization` which flatten all of the actions in the page, so we cannot use ui actions directly.
image

File: src/plugins/discover/public/application/view_components/canvas/discover_canvas.scss [hunk 3]
```diff
@@ -12,6 +12,12 @@
     position: relative;
   }
 
+  &_actions {
+    position: absolute;
+    top: 0;
+    right: 30px;
```
Comment (AMoo-Miki) [inline src/plugins/discover/public/application/view_components/canvas/discover_canvas.scss:18]: How did we land at 30px? We should use $euiSizeL or XL.

**Comment by AMoo-Miki** [comment]
> Yeah, we have considered that solution but it doesn't meet our requirement, ui actions only supports a menu button with multiple actions, the style is different from the UI mockup for generate anomaly detector and text to visualization which flatten all of the actions in the page, so we cannot use ui actions directly.

If i understand correctly, all you need is a different appearance. What are your requirements specifically that are not supported by the existing system?

File: src/plugins/discover/public/application/view_components/canvas/discover_canvas.scss (hunk 3 above)
Comment (gaobinlong) [reply src/plugins/discover/public/application/view_components/canvas/discover_canvas.scss]: Changed it to $euiSizeL, thanks!

**Comment by gaobinlong** [comment]
> > Yeah, we have considered that solution but it doesn't meet our requirement, ui actions only supports a menu button with multiple actions, the style is different from the UI mockup for generate anomaly detector and text to visualization which flatten all of the actions in the page, so we cannot use ui actions directly.
> 
> If i understand correctly, all you need is a different appearance. What are your requirements specifically that are not supported by the existing system?

I think that's because UI actions needs all registered actions behind a clickable element, when clicking that element, all actions are shown, something like a popover,  this is different from our requirement that no clickable entry point exists, but show all the registered actions in the page directly. Maybe we can change the existing UI actions to make it support our case, but the effort is much more than the implementation in this PR.

**Comment by kavilla** [comment]
this seems a little bit of an anti pattern.

Discover has a set of registered actions that it can do into the top nav bar
Screenshot 2024-05-22 at 2 32 16 PM

the other thing is that the histogram is apart of the discover plugin if and only if there is date histogram aggregation which only occurs if timestamps exist. Logically it makes sense for AD that if no timestamp then the action doesn't really matter since I doubt it is possible to do AD without timestamps but for any other type of action this will will not be preset if no timestamp field present.

i think if we really wanted something we could thing bigger and apply it with the query editor component. if you run `yarn start:enhancements` you will get the future iteration of the component.

@joshuali925 has made it really easy to portal in a component from the AD dashboards plugin to the query editor header bar.

**Comment by joshuali925** [comment]
yeah I'm not sure if anomaly detector button should go into the query editor extensions UI, but from user perspective i feel it's a bit weird to have AD next to histogram, since actions are generally on top. It might break the experience if some plugins put actions on top and some put them below

**Comment by dagneyb** [comment]
My two cents: I dont think this action belongs with the top nav "high-level" actions. The call to actions we are putting in line are truly intended to be in in-line with the associated data. We shouldn't let our current architecture drive our design. 

I'm not opposed to putting these in the query editor component as @kavilla called out, but I think we may be up against the deadline to make that change for 2.17.

**Comment by ruanyl** [comment]
Closing as this is not needed
